File: Java/Arrays/ArrayReversal.java
/*
Explanation:
1. We start by defining the `reverseArray` method that takes an array `arr` as input and reverses its elements in-place.
2. We use two pointers, `start` and `end`, pointing to the beginning and end of the array respectively.
3. In each iteration, we swap the elements at `arr[start]` and `arr[end]`.
4. After swapping, we move `start` one step forward and `end` one step backward.
5. We repeat this process until `start` becomes greater than or equal to `end`.
6. In the `main` method, we define a sample input array `arr` and print its original state.
7. We then call the `reverseArray` method to reverse the array in-place.
8. Finally, we print the reversed array.

Time Complexity: The time complexity of this program is O(n), where n is the number of elements in the array. This is because we need to iterate through half of the array to perform the swapping operation.

Space Complexity: The space complexity of this program is O(1) since we are only using a constant amount of extra space to store variables `start`, `end`, and `temp`.

Sample Input:
Original array: [1, 2, 3, 4, 5]

Sample Output:
Reversed array: [5, 4, 3, 2, 1]
*/


import java.util.Arrays;

public class ArrayReversal {
    public static void reverseArray(int[] arr) {
        int start = 0;
        int end = arr.length - 1;
        
        // Swap elements from start to end until both pointers meet in the middle
        while (start < end) {
            // Swap arr[start] and arr[end]
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            
            // Move to the next elements
            start++;
            end--;
        }
    }
    
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5};
        
        System.out.println("Original array: " + Arrays.toString(arr));
        reverseArray(arr);
        
        System.out.println("Reversed array: " + Arrays.toString(arr));
    }
}

File: C++/Arrays/ArrayReversal.cpp
/*
Explanation:
- We define a function `reverseArray` that takes an array and its size as input. The function uses two pointers, `start` and `end`, initialized to the first and last elements of the array, respectively.
- We use a `while` loop that runs until the `start` pointer is less than the `end` pointer. In each iteration, we swap the elements at the `start` and `end` positions and then move both pointers inward towards the center of the array.
- The main function contains a sample input array, prints the original array, calls the `reverseArray` function, and then prints the reversed array.

Sample Input:
{1, 2, 3, 4, 5}

Sample Output:
Original Array: 1 2 3 4 5 
Reversed Array: 5 4 3 2 1 

Time Complexity: The `reverseArray` function runs in linear time O(N), where N is the size of the array, since we traverse half of the array to perform the swapping.
Space Complexity: The space complexity is O(1) since the algorithm only uses a constant amount of extra memory for the `start`, `end`, and `temp` variables, regardless of the array size.
*/

#include <iostream>
// Function to reverse the array
void reverseArray(int arr[], int size) {
    int start = 0;
    int end = size - 1;

    while (start < end) {
        // Swap elements at start and end positions
        int temp = arr[start];
        arr[start] = arr[end];
        arr[end] = temp;

        // Move the start and end pointers inward
        start++;
        end--;
    }
}

int main() {
    // Sample input
    int arr[] = {1, 2, 3, 4, 5};
    int size = sizeof(arr) / sizeof(arr[0]);

    // Print the original array
    std::cout << "Original Array: ";
    for (int i = 0; i < size; i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl;

    // Reverse the array
    reverseArray(arr, size);

    // Print the reversed array
    std::cout << "Reversed Array: ";
    for (int i = 0; i < size; i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}

